### PR TITLE
Use evaluate_function for CallStack::PackageRun; Sample response for CallStack::CheckDeployment

### PR DESCRIPTION
**Pull request by vicsn**
## Motivation

If we `evaluate_function` during deployment, we risk hitting false positive failed constraints on our random assignment.

So instead, this PR proposes to sample a response. After sampling, construction of Response is handled the same way evaluate_function.

## Test Plan

In the future we will test a wider range of programs.

## Related PRs

https://github.com/AleoHQ/snarkVM/pull/2176
https://github.com/AleoHQ/snarkVM/pull/2224
https://github.com/AleoHQ/snarkVM/pull/2226

### CODE REVIEW

File: synthesizer/process/src/stack/helpers/sample.rs [hunk 1]
```diff
@@ -64,6 +44,16 @@ impl<N: Network> Stack<N> {
...
         Ok(plaintext)
     }
+
+    /// Samples a future value according to the given future type.
+    pub fn sample_future<R: Rng + CryptoRng>(&self, locator: &Locator<N>, rng: &mut R) -> Result<Future<N>> {
+        // Sample a future value.
+        let future = self.sample_future_internal(locator, 0, rng)?;
+        // Ensure the plaintext value matches the plaintext type.
+        self.matches_future(&future, locator)?;
+        // Return the plaintext value.
```
Comment (vicsn) [inline synthesizer/process/src/stack/helpers/sample.rs]:
```suggestion
        // Ensure the future value matches the future type.
        self.matches_future(&future, locator)?;
        // Return the future value.
```

File: synthesizer/process/src/stack/helpers/sample.rs [hunk 2]
```diff
@@ -182,4 +172,46 @@ impl<N: Network> Stack<N> {
         // Return the plaintext.
         Ok(plaintext)
     }
+
+    /// Returns a future for the given TODO.
```
Comment (d0cd) [inline synthesizer/process/src/stack/helpers/sample.rs]: Should the `TODO` be swapped out?

File: synthesizer/process/src/stack/call/mod.rs [hunk 3]
```diff
@@ -258,7 +269,65 @@ impl<N: Network> CallTrait<N> for Call<N> {
...
+                        let address = Address::try_from(&private_key)?;
+                        // Sample dummy outputs
+                        let outputs = function
+                            .outputs()
+                            .iter()
+                            .map(|output| match output.value_type() {
+                                ExternalRecord(locator) => {
+                                    // Retrieve the external stack.
+                                    let stack = substack.get_external_stack(locator.program_id())?;
+                                    // Sample the input.
```
Comment (d0cd) [inline synthesizer/process/src/stack/call/mod.rs]:
```suggestion
                                    // Sample the output.
```

File: synthesizer/program/src/traits/stack_and_registers.rs [hunk 4]
```diff
@@ -83,6 +84,14 @@ pub trait StackProgram<N: Network> {
 
     /// Returns the expected number of calls for the given function name.
     fn get_number_of_calls(&self, function_name: &Identifier<N>) -> Result<usize>;
+
+    /// Samples a value for the given value_type
```
Comment (d0cd) [inline synthesizer/program/src/traits/stack_and_registers.rs]:
```suggestion
    /// Samples a value for the given value_type.
```

File: synthesizer/process/src/stack/helpers/matches.rs [hunk 5]
```diff
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 use super::*;
-use console::program::{Argument, FinalizeType};
+pub use console::program::{Argument, FinalizeType};
```
Comment (d0cd) [inline synthesizer/process/src/stack/helpers/matches.rs]: Nit: You can move the `use` here and in `sample.rs` to the module above. It'll let you get rid of the `pub`

**Review by d0cd — APPROVED** [review]
LGTM! Left a few minor comments.

File: synthesizer/process/src/stack/helpers/matches.rs (hunk 5 above)
Comment (vicsn) [reply synthesizer/process/src/stack/helpers/matches.rs]: Thx [7cba76f](https://github.com/AleoHQ/snarkVM/pull/2230/commits/7cba76f6d6f756c636812f336e1448efcb3c7d2c)!

File: synthesizer/program/src/traits/stack_and_registers.rs (hunk 4 above)
Comment (vicsn) [reply synthesizer/program/src/traits/stack_and_registers.rs]: Thx [7cba76f](https://github.com/AleoHQ/snarkVM/pull/2230/commits/7cba76f6d6f756c636812f336e1448efcb3c7d2c)!

File: synthesizer/process/src/stack/call/mod.rs (hunk 3 above)
Comment (vicsn) [reply synthesizer/process/src/stack/call/mod.rs]: Thx [7cba76f](https://github.com/AleoHQ/snarkVM/pull/2230/commits/7cba76f6d6f756c636812f336e1448efcb3c7d2c)!

File: synthesizer/process/src/stack/helpers/sample.rs (hunk 2 above)
Comment (vicsn) [reply synthesizer/process/src/stack/helpers/sample.rs]: Thx [7cba76f](https://github.com/AleoHQ/snarkVM/pull/2230/commits/7cba76f6d6f756c636812f336e1448efcb3c7d2c)!

**Comment by vicsn** [comment]
Did another walk through the code, running on a local devnet and deploying and executing max nested programs succeeds.

File: synthesizer/process/src/stack/call/mod.rs [hunk 6]
```diff
@@ -258,7 +269,65 @@ impl<N: Network> CallTrait<N> for Call<N> {
                         call_stack.push(request.clone())?;
 
                         // Execute the request.
```
Comment (howardwu) [inline synthesizer/process/src/stack/call/mod.rs]:
```suggestion
                        // Evaluate the request.
```

File: synthesizer/process/src/stack/call/mod.rs [hunk 7]
```diff
@@ -257,8 +268,66 @@ impl<N: Network> CallTrait<N> for Call<N> {
...
+                        let outputs = function
+                            .outputs()
+                            .iter()
+                            .map(|output| match output.value_type() {
+                                ExternalRecord(locator) => {
+                                    // Retrieve the external stack.
+                                    let stack = substack.get_external_stack(locator.program_id())?;
+                                    // Sample the output.
+                                    stack.sample_value(&address, &Record(*locator.resource()), rng)
+                                }
```
Comment (howardwu) [inline synthesizer/process/src/stack/call/mod.rs]: Why not move this case directly into the impl of `fn sample_value`?

File: synthesizer/process/src/stack/call/mod.rs [hunk 8]
```diff
@@ -257,8 +268,66 @@ impl<N: Network> CallTrait<N> for Call<N> {
...
+                                _ => substack.sample_value(&address, output.value_type(), rng),
+                            })
+                            .collect::<Result<Vec<_>>>()?;
+
+                        // Retrieve the output operands.
+                        let output_operands =
+                            &function.outputs().iter().map(|output| output.operand()).collect::<Vec<_>>();
```
Comment (howardwu) [inline synthesizer/process/src/stack/call/mod.rs]: Why are we collecting?

File: synthesizer/process/src/stack/call/mod.rs (hunk 7 above)
Comment (vicsn) [reply synthesizer/process/src/stack/call/mod.rs]: Smart a9f348306

File: synthesizer/process/src/stack/call/mod.rs (hunk 8 above)
Comment (vicsn) [reply synthesizer/process/src/stack/call/mod.rs]: Blind code copy pasta 53c69a83e

**Review by howardwu — APPROVED** [review]

**Pull request merged**